### PR TITLE
Refactor JavaTree children() method to return List instead of Iterable

### PR DESCRIPTION
**Pull request by margarita-nedzelska-sonarsource**
Only last commit is relevant

### CODE REVIEW

File: java-frontend/src/main/java/org/sonar/java/ast/parser/ListTreeImpl.java [hunk 1]
```diff
@@ -67,45 +64,8 @@ public Kind kind() {
...
-      iterators = Arrays.asList(list.iterator(), separators.iterator());
-    }
-
-    @Override
-    public Iterator<Tree> iterator() {
-      return new InterleaveIterator<>(iterators);
-    }
-  }
-
-  private static class InterleaveIterator<E> extends AbstractIterator<E> {
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/ast/parser/ListTreeImpl.java:88]: so cool we can drop the AbstractIterator!

File: java-frontend/src/main/java/org/sonar/java/model/JavaTree.java [hunk 2]
```diff
@@ -129,7 +129,7 @@ public void setParent(Tree parent) {
    *
    * @throws java.lang.UnsupportedOperationException if {@link #isLeaf()} returns {@code true}
    */
-  protected abstract Iterable<Tree> children();
+  protected abstract List<Tree> children();
 
   public List<Tree> getChildren() {
     if(children == null) {
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/model/JavaTree.java:135]: So we cab probably change next line to: 
```java
children = children().stream()
  .filter(Objects::nonNull)
  .collect(Collectors.toList());

File: java-frontend/src/main/java/org/sonar/java/model/JavaTree.java [hunk 3]
```diff
@@ -129,7 +129,7 @@ public void setParent(Tree parent) {
    *
    * @throws java.lang.UnsupportedOperationException if {@link #isLeaf()} returns {@code true}
    */
-  protected abstract Iterable<Tree> children();
+  protected abstract List<Tree> children();
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/model/JavaTree.java:132]: There are some unit tests that add the result of this method into another list to do assertions. It worth modifying them to simplify a bit.

**Review by m-g-sonar — CHANGES_REQUESTED** [review]

File: java-frontend/src/main/java/org/sonar/java/ast/parser/ListTreeImpl.java (hunk 1 above)
Comment (margarita-nedzelska-sonarsource) [reply java-frontend/src/main/java/org/sonar/java/ast/parser/ListTreeImpl.java:88]: I guess this class (fully covered btw :) ) lived for a couple of hours

**Review by m-g-sonar — APPROVED** [review]

**Pull request merged**